---
title: "Introduction to occumb"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to occumb}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  comment = "#>"
)
```

```{r, echo = FALSE, message = FALSE}
library(occumb)
set.seed(1)
```

## Dataset

We will use the package's built-in data `fish` (see `?fish` for documentation) to see how to analyze sequence count data with the `occumb` package.
The `summary()` function can be used to display an overview of the dataset.

```{r, eval = FALSE}
data(fish)
summary(fish)
```

```{r}
#> Sequence read counts: 
#> Number of species, I = 50 
#> Number of sites, J = 50 
#> Maximum number of replicates per site, K = 3 
#> Number of missing observations = 6 
#> Number of replicates per site: 2.88 (average), 0.3282607 (sd) 
#> Sequencing depth: 77910.03 (average), 98034.66 (sd) 
#> 
#> Species covariates: 
#>  mismatch (continuous) 
#> Site covariates: 
#>  riverbank (categorical) 
#> Replicate covariates: 
#>  (None) 
#> 
#> Labels for species: 
#>  Abbottina rivularis, Acanthogobius lactipes, Acheilognathus macropterus, Acheilognathus rhombeus, Anguilla japonica, Biwia zezera, Carassius cuvieri, Carassius spp., Channa argus, Ctenopharyngodon idella, Cyprinus carpio, Gambusia affinis, Gnathopogon spp., Gymnogobius castaneus, Gymnogobius petschiliensis, Gymnogobius urotaenia, Hemibarbus spp., Hypomesus nipponensis, Hypophthalmichthys spp., Hyporhamphus intermedius, Ictalurus punctatus, Ischikauia steenackeri, Lepomis macrochirus macrochirus, Leucopsarion petersii, Megalobrama amblycephala, Micropterus dolomieu dolomieu, Micropterus salmoides, Misgurnus spp., Monopterus albus, Mugil cephalus cephalus, Mylopharyngodon piceus, Nipponocypris sieboldii, Nipponocypris temminckii, Opsariichthys platypus, Opsariichthys uncirostris uncirostris, Oryzias latipes, Plecoglossus altivelis altivelis, Pseudogobio spp., Pseudorasbora parva, Rhinogobius spp., Rhodeus ocellatus ocellatus, Salangichthys microdon, Sarcocheilichthys variegatus microoculus, Silurus asotus, Squalidus chankaensis biwae, Tachysurus tokiensis, Tanakia lanceolata, Tribolodon brandtii maruta, Tribolodon hakonensis, Tridentiger spp. 
#> Labels for sites: 
#>  1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50 
#> Labels for replicates: 
#>  L, C, R 
```

Note that a summary of the sequence read count data is found in the first block of the output.
It shows that the data was obtained from samples of 50 sites * (up to) 3 replicates, with 50 fish species recorded.
A summary of the missing samples, number of replicates per site, and sequencing depth (i.e., the total number of sequence reads per sample) are also presented.
The second block of the output indicates that `fish` data also has two covariates, one continuous species covariate, `mismatch`, and one discrete site covariate, `riverbank`, which can be used in the following analysis.
The third block of the output shows labels given to the 50 species, 50 sites, and three replicates.

As such, the `fish` data brings together a set of data relevant to the analysis with the `occumb` package.
When you are analyzing your own dataset, you can use `occumbData()` to set up your data object.

If you want to display the entire data, just type `fish`.

```{r, eval = FALSE}
fish
```

## Model fitting

With the `fish` dataset (or if your dataset is constructed with `occumbData()`), you can fit a multispecies site occupancy model for eDNA metabarcoding using the `occumb()` function.

The `occumb()` function has several arguments that accept model formulas to fit different model variants using the usual R formula syntax. See `vignette("model_specification")` for an overview of the models that `occumb()` can fit and details on how to specify models.

Fitting a null model (i.e., an intercept-only model, though the value of the intercept is species-specific) is very simple, as follows.

```{r, eval = FALSE}
fit0 <- occumb(data = fish, parallel = TRUE)
```

Setting `parallel = TRUE` is recommended for faster model fitting through parallel computation. A model incorporating species and site covariates in the `fish` dataset can be fitted, for example, as follows.

```{r, eval = FALSE}
fit1 <- occumb(formula_psi = ~ riverbank,
               formula_phi_shared = ~ mismatch,
               data = fish,
               parallel = TRUE)
```

Here, `riverbank` is specified as the covariate for site occupancy probability ($\psi$) in the `formula_psi` argument and `mismatch` as the covariate for relative dominance of sequence ($\phi$) in the `formula_phi_shared` argument. In the latter, since `mismatch` is a species covariate, the `formula_phi_shared` argument is used instead of `formula_phi` so that the effect of `mismatch` is constant across species; see `vignette("model_specification")` for more details.

`occumb()` also has several arguments to control the Markov Chain Monte Carlo (MCMC) computation for fitting the model. For example, to obtain more precise posterior estimates than the example above, you can explicitly set the `n.iter` and `n.thin` arguments to get longer and less autocorrelated MCMC samples.

```{r, eval = FALSE}
fit1x <- occumb(formula_psi = ~ riverbank,
                formula_phi_shared = ~ mismatch,
                data = fish,
                n.thin = 20,
                n.iter = 40000,
                parallel = TRUE)
```


